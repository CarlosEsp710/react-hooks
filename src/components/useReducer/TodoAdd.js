import React from "react";

import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAdd }) => {
  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });

  const handleSubmit = (event) => {
    event.preventDefault();

    if (description.trim().length <= 2) {
      return;
    }

    handleAdd({
      id: new Date().getTime(),
      desc: description,
      done: false,
    });

    reset();
  };

  return (
    <>
      <h4>Add Todo</h4>
      <hr />
      <form onSubmit={handleSubmit} className="d-grid gap-2">
        <input
          type="text"
          name="description"
          placeholder="Write..."
          autoComplete="off"
          className="form-control"
          value={description}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-sm btn-block"
        >
          Add
        </button>
      </form>
    </>
  );
};
