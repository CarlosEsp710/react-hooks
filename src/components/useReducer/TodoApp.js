import React, { useEffect, useReducer } from "react";

import { todoReducer } from "./todoReducer";
import { useForm } from "../../hooks/useForm";

import "./reducer.css";

const init = () => JSON.parse(localStorage.getItem("todos")) || [];

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (id) => {
    const action = {
      type: "delete",
      payload: id,
    };

    dispatch(action);
  };

  const handleToggle = (id) => {
    dispatch({
      type: "toggle",
      payload: id,
    });
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    const action = {
      type: "add",
      payload: newTodo,
    };

    dispatch(action);

    reset();
  };

  return (
    <>
      <h1>TodoApp ({todos.length})</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <ul className="list-group list-group-flush">
            {todos.map((todo, i) => (
              <li key={todo.id} className="list-group-item">
                <p
                  onClick={() => handleToggle(todo.id)}
                  className={`${todo.done && "complete"}`}
                >
                  {i + 1}. {todo.desc}
                </p>
                <button
                  onClick={() => handleDelete(todo.id)}
                  className="btn btn-danger"
                >
                  Delete
                </button>
              </li>
            ))}
          </ul>
        </div>
        <div className="col-5">
          <h4>Add Todo</h4>
          <hr />
          <form onSubmit={handleSubmit} className="d-grid gap-2">
            <input
              type="text"
              name="description"
              placeholder="Write..."
              autoComplete="off"
              className="form-control"
              value={description}
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn btn-outline-primary mt-1 btn-sm btn-block"
            >
              Add
            </button>
          </form>
        </div>
      </div>
    </>
  );
};
