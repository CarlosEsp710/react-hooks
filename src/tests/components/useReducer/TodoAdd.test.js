import React from "react";
import { shallow } from "enzyme";

import { TodoAdd } from "../../../components/useReducer/TodoAdd";

describe("test <TodoAdd />", () => {
  const handleAdd = jest.fn();

  const wrapper = shallow(<TodoAdd handleAdd={handleAdd} />);

  test("should return component <TodoAdd />", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should not call handleAdd if value is empty", () => {
    wrapper.find("button").simulate("submit", { preventDefault() {} });

    expect(handleAdd).not.toHaveBeenCalled();
  });

  test("should call handleAdd if value is not empty", () => {
    const value = "Aprender Java";

    const input = wrapper
      .find("input")
      .simulate("change", { target: { value, name: "description" } });

    wrapper.find("form").simulate("submit", { preventDefault() {} });

    expect(handleAdd).toHaveBeenCalledTimes(1);
    expect(handleAdd).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAdd).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false,
    });
    expect(input.prop("value")).toBe("");
  });
});
