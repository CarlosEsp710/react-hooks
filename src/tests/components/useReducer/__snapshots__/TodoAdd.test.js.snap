// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <TodoAdd /> should return component <TodoAdd /> 1`] = `
<Fragment>
  <h4>
    Add Todo
  </h4>
  <hr />
  <form
    className="d-grid gap-2"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Write..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-sm btn-block"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
